File: apps/api/app/binary.py
"""Pack/unpack helpers for the exact little-endian binary buffer layouts
described in docs/API_CONTRACT.md. All arrays are little-endian float32
(`<f4`) or uint32 (`<u4`), produced via numpy so the byte layout is exact
and portable regardless of host endianness.
"""

import numpy as np

F32 = "<f4"
U32 = "<u4"
I32 = "<i4"


def pack_well_log(depths: np.ndarray, values: np.ndarray) -> bytes:
    """Flat Float32Array of interleaved [depth0, value0, depth1, value1, ...]."""
    depths = np.asarray(depths, dtype=F32)
    values = np.asarray(values, dtype=F32)
    interleaved = np.empty(depths.size + values.size, dtype=F32)
    interleaved[0::2] = depths
    interleaved[1::2] = values
    return interleaved.tobytes()


def unpack_well_log(data: bytes) -> tuple[np.ndarray, np.ndarray]:
    arr = np.frombuffer(data, dtype=F32)
    return arr[0::2].copy(), arr[1::2].copy()


def pack_grid_cells(centers: np.ndarray, sizes: np.ndarray, cell_ids: np.ndarray) -> bytes:
    """centers, sizes: (cellCount, 3) float arrays. cell_ids: (cellCount,) int array.

    Layout: Float32Array[cellCount*3] centers, Float32Array[cellCount*3] sizes,
    Uint32Array[cellCount] cellId.
    """
    centers = np.ascontiguousarray(centers, dtype=F32)
    sizes = np.ascontiguousarray(sizes, dtype=F32)
    cell_ids = np.ascontiguousarray(cell_ids, dtype=U32)
    return centers.tobytes() + sizes.tobytes() + cell_ids.tobytes()


def pack_float32_array(values: np.ndarray) -> bytes:
    return np.ascontiguousarray(values, dtype=F32).tobytes()


def unpack_float32_array(data: bytes) -> np.ndarray:
    return np.frombuffer(data, dtype=F32).copy()


def pack_ijk(ijk: np.ndarray) -> bytes:
    """(cellCount, 3) int array of (i, j, k) natural-order indices per active
    cell. Internal-only: not part of the public API contract, used solely by
    the /sample endpoint to reconstruct the structured lattice server-side."""
    return np.ascontiguousarray(ijk, dtype=I32).tobytes()


def unpack_ijk(data: bytes) -> np.ndarray:
    return np.frombuffer(data, dtype=I32).copy().reshape(-1, 3)

File: apps/web/lib/binary.ts
// Binary decode helpers matching the layouts specified in docs/API_CONTRACT.md
// ("Grids", "Wells", "Surfaces" sections). All buffers are little-endian, which
// is what Float32Array/Uint32Array use natively on all browser platforms (they
// read/write in the platform's native byte order, and every platform we target
// -- x86/ARM browsers -- is little-endian), so no DataView byte-swapping is
// required as long as the buffer is not itself a sub-view with a non-zero,
// non-4-byte-aligned offset. We defensively copy via `.slice()` when the
// incoming ArrayBuffer's byteOffset could break alignment (fetch() always
// returns a fresh, zero-offset ArrayBuffer, but we guard anyway).

export interface DecodedGridCells {
  cellCount: number;
  /** Float32Array[cellCount*3], interleaved (x,y,z) per instance */
  centers: Float32Array;
  /** Float32Array[cellCount*3], interleaved (dx,dy,dz) per instance */
  sizes: Float32Array;
  /** Uint32Array[cellCount], index into the property arrays at the same lod */
  cellIds: Uint32Array;
}

/**
 * Decode the /grids/{id}/cells binary payload.
 *
 * Layout (little-endian), all arrays length `cellCount`:
 *  - Float32Array[cellCount*3] -- instance centers (x,y,z)
 *  - Float32Array[cellCount*3] -- instance sizes (dx,dy,dz)
 *  - Uint32Array[cellCount]    -- cellId
 *
 * `cellCount` is derivable from Content-Length (or passed in explicitly from
 * the `X-Cell-Count` response header, which is the authoritative source).
 */
export function decodeGridCells(buf: ArrayBuffer, cellCountHint?: number): DecodedGridCells {
  // Each cell contributes: 3 floats (center) + 3 floats (size) + 1 uint32 (id)
  // = (3*4 + 3*4 + 4) = 28 bytes/cell.
  const bytesPerCell = 3 * 4 + 3 * 4 + 4;
  const cellCount = cellCountHint ?? Math.floor(buf.byteLength / bytesPerCell);

  const centersBytes = cellCount * 3 * 4;
  const sizesBytes = cellCount * 3 * 4;

  const centers = new Float32Array(buf.slice(0, centersBytes));
  const sizes = new Float32Array(buf.slice(centersBytes, centersBytes + sizesBytes));
  const cellIds = new Uint32Array(buf.slice(centersBytes + sizesBytes, centersBytes + sizesBytes + cellCount * 4));

  return { cellCount, centers, sizes, cellIds };
}

/**
 * Decode /grids/{id}/properties/{name} -- a flat Float32Array[cellCount], one
 * value per cell, in the same order/subsampling as /cells at that lod.
 */
export function decodeGridProperty(buf: ArrayBuffer): Float32Array {
  return new Float32Array(buf.slice(0));
}

/**
 * Decode /surfaces/{id}/heights -- Float32Array[rows*cols], row-major, NaN
 * where nodata. `rows`/`cols` come from the SurfaceDetail and are only used
 * here for validation/reshaping convenience.
 */
export function decodeSurfaceHeights(buf: ArrayBuffer, rows: number, cols: number): Float32Array {
  const heights = new Float32Array(buf.slice(0));
  const expected = rows * cols;
  if (heights.length !== expected) {
    // Don't throw -- surface may still be partially usable; just warn.
    // eslint-disable-next-line no-console
    console.warn(
      `decodeSurfaceHeights: expected ${expected} values (rows=${rows} * cols=${cols}), got ${heights.length}`
    );
  }
  return heights;
}

export interface DecodedLogData {
  /** sample count (number of depth/value pairs) */
  count: number;
  depth: Float32Array;
  value: Float32Array;
}

/**
 * Decode /wells/{id}/logs/{log_id}/data -- a flat Float32Array of interleaved
 * [depth0, value0, depth1, value1, ...], little-endian.
 * `buffer.byteLength === sample_count * 2 * 4`.
 */
export function decodeLogData(buf: ArrayBuffer, sampleCountHint?: number): DecodedLogData {
  const interleaved = new Float32Array(buf.slice(0));
  const count = sampleCountHint ?? Math.floor(interleaved.length / 2);
  const depth = new Float32Array(count);
  const value = new Float32Array(count);
  for (let i = 0; i < count; i++) {
    depth[i] = interleaved[i * 2];
    value[i] = interleaved[i * 2 + 1];
  }
  return { count, depth, value };
}
